File: mygreeterv3csharp/server/Src/Client/Operation/Start.cs
using System.CommandLine;
using System.CommandLine.NamingConventionBinder;
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;
using MyGreeterCsharp.Api.V1;
using Serilog;
using Serilog.Templates;

namespace MyGreeterCsharp.Client.Operation;

public class ClientOptions
{
    public string? RemoteAddr { get; set; }
    public string? HttpAddr { get; set; }
    public bool JsonLog { get; set; }
    public string? Name { get; set; }
    public int Age { get; set; }
    public string? Email { get; set; }
    public string? Address { get; set; }
    public long IntervalMilliSec { get; set; }
    public string? RgName { get; set; }
    public string? RgRegion { get; set; }
    public bool CallAllRgOps { get; set; }
}


public static class StartCommand
{
    public static Command Init()
    {
        var remoteAddrOption = new Option<string>(
            "--remote-addr",
            description: "The remote server's address for this client to connect to",
            getDefaultValue: () => "localhost:50051");

        var httpAddrOption = new Option<string>(
            "--http-addr",
            description: "The remote HTTP gateway address",
            getDefaultValue: () => "http://localhost:50061");

        var jsonLogOption = new Option<bool>(
            "--json-log",
            description: "Enables JSON format for logs (human readable key-value pairs)",
            getDefaultValue: () => false);

        var nameOption = new Option<string>(
            "--name",
            description: "The name to send in Hello request",
            getDefaultValue: () => "MyName");

        var ageOption = new Option<int>(
            "--age",
            description: "The age to send in Hello request",
            getDefaultValue: () => 53);

        var emailOption = new Option<string>(
            "--email",
            description: "The email to send in Hello request",
            getDefaultValue: () => "test@test.com");

        var addressOption = new Option<string>(
            "--address",
            description: "The address to send in Hello request",
            getDefaultValue: () => "123 Main St, Seattle, WA 98101");

        var intervalMilliSecOption = new Option<long>(
            "--interval-milli-sec",
            description: "The interval between two requests. Negative numbers mean sending one request.",
            getDefaultValue: () => -1);

        var rgNameOption = new Option<string>(
            "--rg-name",
            description: "The name of the resource group",
            getDefaultValue: () => "MyGreeterCsharp-resource-group");

        var rgRegionOption = new Option<string>(
            "--rg-region",
            description: "The region of the resource group",
            getDefaultValue: () => "eastus");

        var callAllRgOpsOption = new Option<bool>(
            "--call-all-rg-ops",
            description: "Call all resource group operations",
            getDefaultValue: () => true);

        var startCommand = new Command("hello", "Call SayHello")
        {
            remoteAddrOption,
            httpAddrOption,
            jsonLogOption,
            nameOption,
            ageOption,
            emailOption,
            addressOption,
            intervalMilliSecOption,
            rgNameOption,
            rgRegionOption,
            callAllRgOpsOption
        };

        startCommand.Handler = CommandHandler.Create<ClientOptions>(hello);

        return startCommand;
    }

    // hello is a client function that configures logging, creates a new client, and calls the SayHello function
    public static async Task hello(ClientOptions options)
    {

        // Serilog configuration
        var loggerConfiguration = new LoggerConfiguration()
            .Enrich.FromLogContext()
            .Enrich.With<LogAttrs.CustomAttributeEnricher>();

        if (options.JsonLog)
        {
            loggerConfiguration = loggerConfiguration.WriteTo.Console(new ExpressionTemplate(
                "{ {time: @t, level: if @l = 'Information' then 'INFO' else if @l = 'Error' then 'ERROR' else if @l = 'Warning' then 'WARN' else if @l = 'Debug' then 'DEBUG' else if @l = 'Verbose' then 'VERBOSE' else if @l = 'Fatal' then 'FATAL' else @l, msg: @m, EX: @x, location: @Location, ..@p} }\n"));
        }
        else
        {
            loggerConfiguration = loggerConfiguration.WriteTo.Console(outputTemplate: "{Timestamp} [{Level}] {Message} {CustomAttributes:lj}{Properties}{NewLine}{Exception}");
        }

        Log.Logger = loggerConfiguration.CreateLogger();

        var client = ClientFactory.NewClient(options.RemoteAddr!, Log.Logger);

        if (options.IntervalMilliSec < 0)
        {
            await SayHello(client, options.Name!, options.Age!, options.Email!, options.Address!, options);
        }
        else
        {
            while (true)
            {
                await SayHello(client, options.Name!, options.Age!, options.Email!, options.Address!, options);
                await Task.Delay((int)options.IntervalMilliSec!);
            }
        }
    }

    /// <summary>
    /// Sends a greeting request to the gRPC server and performs various resource group operations.
    /// </summary>
    /// <param name="client">The gRPC client used to send requests.</param>
    /// <param name="name">The name of the user to greet.</param>
    /// <param name="age">The age of the user to greet.</param>
    /// <param name="email">The email of the user to greet.</param>
    /// <param name="address">The address of the user to greet, in the format "Street, City, State Zip".</param>
    /// <param name="options">The client options containing resource group information.</param>
    /// <returns>A task representing the asynchronous operation.</returns>
    /// <remarks>
    /// This method performs the following operations:
    /// <list type="bullet">
    /// <item><description>Sends a greeting request to the gRPC server.</description></item>
    /// <item><description>Creates a resource group.</description></item>
    /// <item><description>Lists resource groups.</description></item>
    /// <item><description>Reads a resource group.</description></item>
    /// <item><description>Updates a resource group.</description></item>
    /// <item><description>Deletes a resource group.</description></item>
    /// </list>
    /// </remarks>    
    private static async Task SayHello(MyGreeterCsharp.Api.V1.MyGreeterCsharp.MyGreeterCsharpClient client, string name, int age, string email, string address, ClientOptions options)
    {
        if (address.Split(',').Length != 3 || address.Split(',')[2].Trim().Split(' ').Length != 2)
        {
            throw new ArgumentException("Address format is incorrect. Expected format: 'Street, City, State Zip'.");
        }

        string[] addressParts = address.Split(',');
        string street = addressParts[0].Trim();
        string city = addressParts[1].Trim();
        string[] stateAndZip = addressParts[2].Trim().Split(' ');
        string state = stateAndZip[0];
        string zipString = stateAndZip[1];
        if (!int.TryParse(zipString, out int zipCode))
        {
            throw new ArgumentException("Invalid zip code format.");
        }

        var addr = new Address
        {
            Street = street,
            City = city,
            State = state,
            Zipcode = zipCode
        };

        var helloRequest = new HelloRequest
        {
            Name = name,
            Age = age,
            Email = email,
            Address = addr
        };

        try
        {
            var reply = await client.SayHelloAsync(helloRequest);
            Log.Information("Greeting: {Message}", reply.Message);
        }
        catch (RpcException ex)
        {
            Log.Error($"gRPC Error calling SayHello: {ex.Status.Detail}");
        }
        catch (Exception ex)
        {
            Log.Error("Error: {Message}", ex.Message);
        }

        try
        {
            var reply = await client.CreateResourceGroupAsync(new CreateResourceGroupRequest
            {
                Name = options.RgName,
                Region = options.RgRegion
            });
        }
        catch (RpcException ex)
        {
            Log.Error($"gRPC Error calling CreateResourceGroup: {ex.Status.Detail}");
        }
        catch (Exception ex)
        {
            Log.Error("Error: {Message}", ex.Message);
        }

        try
        {
            var response = await client.ListResourceGroupsAsync(new Empty());
        }
        catch (RpcException ex)
        {
            Log.Error($"gRPC Error calling ListResourceGroup: {ex.Status.Detail}");
        }
        catch (Exception ex)
        {
            Log.Error($"Error calling ListResourceGroup: {ex.Message}");
        }

        try
        {
            var response = await client.ReadResourceGroupAsync(new ReadResourceGroupRequest
            {
                Name = options.RgName
            });
        }
        catch (RpcException ex)
        {
            Log.Error($"gRPC Error calling ReadResourceGroup: {ex.Status.Detail}");
        }
        catch (Exception ex)
        {
            Log.Error($"Error calling ReadResourceGroup: {ex.Message}");
        }

        try
        {
            var tags = new Dictionary<string, string>
            {
                { "key1", "value1" },
                { "key2", "value2" }
            };

            var response = await client.UpdateResourceGroupAsync(new UpdateResourceGroupRequest
            {
                Name = options.RgName,
                Tags = { tags }
            });
        }
        catch (RpcException ex)
        {
            Log.Error($"gRPC Error calling UpdateResourceGroup: {ex.Status.Detail}");
        }
        catch (Exception ex)
        {
            Log.Error($"Error calling UpdateResourceGroup: {ex.Message}");
        }

        try
        {
            var response = await client.DeleteResourceGroupAsync(new DeleteResourceGroupRequest
            {
                Name = options.RgName
            });
        }
        catch (RpcException ex)
        {
            Log.Error($"gRPC Error calling DeleteResourceGroup: {ex.Status.Detail}");
        }
        catch (Exception ex)
        {
            Log.Error($"Error calling DeleteResourceGroup: {ex.Message}");
        }

        // TODO: Add rest API calls when RestSDK is implemented 

        // try
        // {
        //     var createReply = await client.CreateStorageAccountAsync(new CreateStorageAccountRequest
        //     {
        //         RgName = options.RgName,
        //         Region = options.RgRegion
        //     });
        // }
        // catch (RpcException ex)
        // {
        //     Log.Error($"gRPC Error calling CreateStorageAccount: {ex.Status.Detail}");
        // }
        // catch (Exception ex)
        // {
        //     Log.Error("Error: {Message}", ex.Message);
        // }

        // try
        // {
        //     var readReply = await client.ReadStorageAccountAsync(new ReadStorageAccountRequest
        //     {
        //         RgName = options.RgName,
        //         SaName = "StorageName"
        //     });
        // }
        // catch (RpcException ex)
        // {
        //     Log.Error($"gRPC Error calling ReadStorageAccount: {ex.Status.Detail}");
        // }
        // catch (Exception ex)
        // {
        //     Log.Error("Error: {Message}", ex.Message);
        // }

        // try
        // {
        //     await client.DeleteStorageAccountAsync(new DeleteStorageAccountRequest
        //     {
        //         RgName = options.RgName,
        //         SaName = "StorageName"
        //     });
        // }
        // catch (RpcException ex)
        // {
        //     Log.Error($"gRPC Error calling DeleteStorageAccount: {ex.Status.Detail}");
        // }
        // catch (Exception ex)
        // {
        //     Log.Error("Error: {Message}", ex.Message);
        // }

        // try
        // {
        //     var updateReply = await client.UpdateStorageAccountAsync(new UpdateStorageAccountRequest
        //     {
        //         RgName = options.RgName,
        //         SaName = "StorageName",
        //         Tags = { { "tag1", "value1" }, { "tag2", "value2" } }
        //     });
        // }
        // catch (RpcException ex)
        // {
        //     Log.Error($"gRPC Error calling UpdateStorageAccount: {ex.Status.Detail}");
        // }
        // catch (Exception ex)
        // {
        //     Log.Error("Error: {Message}", ex.Message);
        // }

        // try
        // {
        //     var listReply = await client.ListStorageAccountsAsync(new ListStorageAccountRequest
        //     {
        //         RgName = options.RgName
        //     });
        // }
        // catch (RpcException ex)
        // {
        //     Log.Error($"gRPC Error calling ListStorageAccounts: {ex.Status.Detail}");
        // }
        // catch (Exception ex)
        // {
        //     Log.Error("Error: {Message}", ex.Message);
        // }
    }
}

File: mygreeterv3/server/cmd/client/start.go
package main

import (
	"context"
	"io"
	"math/rand"
	"os"
	"time"

	"dev.azure.com/service-hub-flg/service_hub_validation/_git/service_hub_validation_service.git/mygreeterv3/api/v1/client"
	"dev.azure.com/service-hub-flg/service_hub_validation/_git/service_hub_validation_service.git/mygreeterv3/server/internal/logattrs"

	"strconv"

	pb "dev.azure.com/service-hub-flg/service_hub_validation/_git/service_hub_validation_service.git/mygreeterv3/api/v1"
	"dev.azure.com/service-hub-flg/service_hub_validation/_git/service_hub_validation_service.git/mygreeterv3/api/v1/restsdk"
	"github.com/Azure/aks-middleware/grpc/interceptor"
	"github.com/Azure/aks-middleware/http/client/direct/restlogger"
	"google.golang.org/protobuf/types/known/timestamppb"

	"strings"

	log "log/slog"

	"github.com/spf13/cobra"
)

var startCmd = &cobra.Command{
	Use:   "hello",
	Short: "Call SayHello",
	Run:   hello,
}

var output io.Writer = os.Stdout

type Options struct {
	RemoteAddr       string
	HttpAddr         string
	JsonLog          bool
	Name             string
	Age              int32
	Email            string
	Address          string
	IntervalMilliSec int64
	RgName           string
	RgRegion         string
	CallAllRGOps     bool
	SubscriptionId   string
}

var options = newOptions()

func init() {
	rootCmd.AddCommand(startCmd)

	startCmd.Flags().StringVar(&options.RemoteAddr, "remote-addr", options.RemoteAddr, "the remote server's addr for this client to connect to")
	startCmd.Flags().StringVar(&options.HttpAddr, "http-addr", options.HttpAddr, "the remote http gateway addr")
	startCmd.Flags().BoolVar(&options.JsonLog, "json-log", options.JsonLog, "The format of the log is json or user friendly key-value pairs")
	startCmd.Flags().StringVar(&options.Name, "name", options.Name, "The name to send in Hello request")
	startCmd.Flags().Int32Var(&options.Age, "age", options.Age, "The age to send in Hello request")
	startCmd.Flags().StringVar(&options.Email, "email", options.Email, "The email to send in Hello request")
	startCmd.Flags().StringVar(&options.Address, "address", options.Address, "The address to send in Hello request")
	startCmd.Flags().Int64Var(&options.IntervalMilliSec, "interval-milli-sec", options.IntervalMilliSec,
		"The interval between two requests. Negative numbers mean sending one request.")
	startCmd.Flags().StringVar(&options.RgName, "rg-name", options.RgName, "The name of the resource group")
	startCmd.Flags().StringVar(&options.RgRegion, "rg-region", options.RgRegion, "The region of the resource group")
	startCmd.Flags().BoolVar(&options.CallAllRGOps, "call-all-rg-ops", options.CallAllRGOps, "Call all resource group operations")
	startCmd.Flags().StringVar(&options.SubscriptionId, "subscription-id", options.SubscriptionId, "The Azure subscription ID to use for API calls")

	// Seed at the beginning of the program once, not on every run: https://stackoverflow.com/questions/12321133/how-to-properly-seed-random-number-generator
	rand.Seed(time.Now().UnixNano())
}

func newOptions() Options {
	return Options{
		RemoteAddr: "localhost:50051",
		// Don't use proxy locally due to audit log validation issues, send REST request directly to grpc gateway port
		HttpAddr:         "http://localhost:50061",
		JsonLog:          false,
		Name:             "MyName",
		Age:              53,
		Email:            "test@test.com",
		Address:          "123 Main St, Seattle, WA 98101",
		IntervalMilliSec: -1,
		RgName:           "MyGreeter-resource-group",
		RgRegion:         "eastus",
		CallAllRGOps:     true,
		SubscriptionId:   "8ecadfc9-d1a3-4ea4-b844-0d9f87e4d7c8",
	}
}

func SetOutput(out io.Writer) {
	output = out
}

func hello(cmd *cobra.Command, args []string) {
	logger := log.New(log.NewTextHandler(output, nil).WithAttrs(logattrs.GetAttrs()))
	if options.JsonLog {
		logger = log.New(log.NewJSONHandler(output, nil).WithAttrs(logattrs.GetAttrs()))
	}

	log.SetDefault(logger)

	client, conn, err := client.NewClient(options.RemoteAddr, interceptor.GetClientInterceptorLogOptions(logger, logattrs.GetAttrs()))
	// logging the error for transparency, retry interceptor will handle it
	if err != nil {
		log.Error("did not connect: " + err.Error())
	}

	if options.IntervalMilliSec < 0 {
		SayHello(client, options.Name, options.Age, options.Email, options.Address, logger)
		conn.Close()
		return
	}
	for {
		SayHello(client, options.Name, options.Age, options.Email, options.Address, logger)
		time.Sleep(time.Duration(options.IntervalMilliSec) * time.Millisecond)

		defer conn.Close()
	}
}

func SayHello(client pb.MyGreeterClient, name string, age int32, email string, address string, logger *log.Logger) {
	ctx := context.Background()

	addressParts := strings.Split(address, ",")
	street := addressParts[0]
	city := addressParts[1]
	stateAndZip := strings.Split(addressParts[2], " ")
	state := stateAndZip[1]
	parseZipcode, _ := strconv.ParseInt(stateAndZip[2], 10, 32)
	zipcode := int32(parseZipcode)

	addr := &pb.Address{
		Street:  street,
		City:    city,
		State:   state,
		Zipcode: zipcode,
	}
	r, err := client.SayHello(ctx, &pb.HelloRequest{Name: name, Age: age, Email: email, Address: addr})
	if err != nil {
		log.Info("SayHello error: " + err.Error())
	} else {
		log.Info("Response message: " + r.GetMessage())
	}

	tags := map[string]string{
		"tag1": "value1",
		"tag2": "value2",
	}

	// Create a new Configuration instance
	cfg := &restsdk.Configuration{
		BasePath: options.HttpAddr,
		DefaultHeader: map[string]string{
			"Content-Type":       "application/json",
			"x-ms-client-app-id": "MyGreeter-Client",
		},
		UserAgent:  "Swagger-Codegen/1.0.0/go",
		HTTPClient: restlogger.NewLoggingClient(logger),
	}

	apiClient := restsdk.NewAPIClient(cfg)

	service := apiClient.MyGreeterApi

	// Include subID in REST request URL to conform with ARM URL formating (/subscriptions/...) for audit logging
	subId := options.SubscriptionId

	// Only update resource groups and perform storage account CRUDL ops
	// in int and stg environments due to current MSI role
	if !options.CallAllRGOps {
		_, err = client.UpdateResourceGroup(ctx, &pb.UpdateResourceGroupRequest{SubscriptionId: subId, Name: options.RgName, Tags: tags})
		if err != nil {
			log.Error("Error calling UpdateResourceGroup: " + err.Error())
		}
		_, _, err = service.MyGreeterUpdateResourceGroup(context.Background(), tags, subId, options.RgName)
		if err != nil {
			log.Error("Error calling MyGreeterUpdateResourceGroup: " + err.Error())
		}
		resp, createSAErr := client.CreateStorageAccount(ctx, &pb.CreateStorageAccountRequest{SubscriptionId: subId, RgName: options.RgName, SaName: "testsa" + strconv.Itoa(rand.Intn(1000)), Region: options.RgRegion})
		if createSAErr != nil {
			log.Error("Error calling CreateStorageAccount: " + err.Error())
		}
		storageAccountName := resp.Name
		_, err = client.ReadStorageAccount(ctx, &pb.ReadStorageAccountRequest{SubscriptionId: subId, RgName: options.RgName, SaName: storageAccountName})
		if err != nil {
			log.Error("Error calling ReadStorageAccount: " + err.Error())
		}
		_, err = client.UpdateStorageAccount(ctx, &pb.UpdateStorageAccountRequest{SubscriptionId: subId, RgName: options.RgName, SaName: storageAccountName, Tags: tags})
		if err != nil {
			log.Error("Error calling UpdateStorageAccount: " + err.Error())
		}
		_, err = client.ListStorageAccounts(ctx, &pb.ListStorageAccountRequest{SubscriptionId: subId, RgName: options.RgName})
		if err != nil {
			log.Error("Error calling ListStorageAccounts: " + err.Error())
		}
		_, err = client.DeleteStorageAccount(ctx, &pb.DeleteStorageAccountRequest{SubscriptionId: subId, RgName: options.RgName, SaName: storageAccountName})
		if err != nil {
			log.Error("Error calling DeleteStorageAccount: " + err.Error())
		}

	} else {
		_, err = client.CreateResourceGroup(ctx, &pb.CreateResourceGroupRequest{SubscriptionId: subId, Name: options.RgName, Region: options.RgRegion})
		if err != nil {
			log.Error("Error calling CreateResourceGroup: " + err.Error())
		}
		_, err = client.ReadResourceGroup(ctx, &pb.ReadResourceGroupRequest{SubscriptionId: subId, Name: options.RgName})
		if err != nil {
			log.Error("Error calling ReadResourceGroup: " + err.Error())
		}
		_, err = client.UpdateResourceGroup(ctx, &pb.UpdateResourceGroupRequest{SubscriptionId: subId, Name: options.RgName, Tags: tags})
		if err != nil {
			log.Error("Error calling UpdateResourceGroup: " + err.Error())
		}
		_, err = client.ListResourceGroups(ctx, &pb.ListResourceGroupsRequest{SubscriptionId: subId})
		if err != nil {
			log.Error("Error calling ListResourceGroup: " + err.Error())
		}
		_, err = client.DeleteResourceGroup(ctx, &pb.DeleteResourceGroupRequest{SubscriptionId: subId, Name: options.RgName})
		if err != nil {
			log.Error("Error calling DeleteResourceGroup: " + err.Error())
		}

		createRequestBody := restsdk.MyGreeterCreateResourceGroupBody{
			Region: options.RgRegion,
		}
		_, _, err = service.MyGreeterCreateResourceGroup(context.Background(), createRequestBody, subId, options.RgName)
		if err != nil {
			log.Error("Error calling MyGreeterCreateResourceGroup: " + err.Error())
		}
		_, _, err = service.MyGreeterReadResourceGroup(context.Background(), subId, options.RgName)
		if err != nil {
			log.Error("Error calling MyGreeterReadResourceGroup: " + err.Error())
		}
		_, _, err = service.MyGreeterUpdateResourceGroup(context.Background(), tags, subId, options.RgName)
		if err != nil {
			log.Error("Error calling MyGreeterUpdateResourceGroup: " + err.Error())
		}
		_, _, err = service.MyGreeterListResourceGroups(context.Background(), subId)
		if err != nil {
			log.Error("Error calling MyGreeterListResourceGroups: " + err.Error())
		}
		_, _, err = service.MyGreeterDeleteResourceGroup(context.Background(), subId, options.RgName)
		if err != nil {
			log.Error("Error calling MyGreeterDeleteResourceGroup: " + err.Error())
		}
	}

	protoExpirationTime := timestamppb.New(time.Now().Add(1 * time.Hour))
	randomNumber := rand.Intn(100) + 1
	randomNumberStr := strconv.Itoa(randomNumber)
	startOperationRequest := &pb.StartLongRunningOperationRequest{
		EntityId:            randomNumberStr,
		EntityType:          "Cluster",
		ExpirationTimestamp: protoExpirationTime,
	}
	startOperationResponse, err := client.StartLongRunningOperation(ctx, startOperationRequest)
	if err != nil {
		log.Error("Error calling StartOperation: " + err.Error())
	} else {
		log.Info("Started operation with operationId: " + startOperationResponse.GetOperationId())
	}
}
